blob/azureblob: Add support for specifying Azure Cloud Environment

Strife96 opened the pull request:
Fixes #2716
Updated 12/24/19 to reflect changes in implementation:
The current go-cloud azureblob implementation always targets "blob.core.windows.net" when attempting to open a bucket, making USGovernment, China, and German cloud buckets inaccessible. 

With this addition, users can specify the Azure Cloud Environment they would like to target by setting  the AZURE_STORAGE_DOMAIN environment variable to a value like "blob.core.govcloudapi.net", "blob.core.chinacloudapi.cn", etc. allowing them to access the different Azure blob storage endpoints for each of these clouds. 

If users leave the AZURE_STORAGE_DOMAIN variable unset, the bucket opener will target "blob.core.windows.net" by default, which is the public domain that most users are likely already using. Any changes made by this request should be transparent to end users who have not been setting this environment variable. 

If you have any questions or feedback regarding this change, please let me know! Thank you! 🙂

vangent commented on blob/azureblob/azureblob.go:
```diff
@@ -96,6 +96,8 @@ type Options struct {
 	// See https://docs.microsoft.com/en-us/azure/storage/common/storage-dotnet-shared-access-signature-part-1#shared-access-signature-parameters.
 	SASToken SASToken
+
+	CloudEnvironment CloudEnvironment
```
Can you add a comment here explaining what this option is, what the possible values are, and what it defaults to?

vangent commented on blob/azureblob/azureblob.go:
```diff
@@ -223,6 +226,18 @@ type AccountKey string
+type CloudEnvironment string
+
+// Mappings from the AZURE_CLOUD_ENVIRONMENT value to their respective blob storage domains.
+var cloudEnvMap = map[CloudEnvironment]string{
```
Is there an advantage to having this mapping? I.e., why not have the CloudEnvironment variable just be the storage domain itself?

For example, with the mapping approach we'll need to update the code every time there's a new environment.

Strife96 replied:
The purpose of the mapping is to be able to use the same environment variable values ("AzureCloud", "AzureUSGovernment", etc.) that users may already be familiar with for this functionality (especially if they use Ansible, which is fairly popular). Furthermore, the values supplied mirror the Azure CLI and Azure Powershell names for these Cloud environments (when you do an "az cloud list" for example). It seemed a little more user-friendly in the end. 

However, if you think this setup is too hard to maintain into the future (I'm unsure how often Azure modifies or adds environments/endpoints), I can easily change the option to simply take in a specific URL suffix instead, it's certainly more flexible. The environment variable name should probably also be changed though (maybe something like GOCLOUD_AZURE_ENVIRONMENT?), so as not to collide with Ansible users who will be using AZURE_CLOUD_ENVIRONMENT with different values. Would you like me to do this instead?

vangent commented on blob/azureblob/azureblob_test.go:
```diff
@@ -53,6 +55,7 @@ import (
 const (
 	bucketName  = "go-cloud-bucket"
 	accountName = AccountName("gocloudblobtests")
+	cloudEnv    = CloudEnvironment("")
```
I think you can leave this out as well.

Strife96 replied:
Sure (also to all the below changes). Something made me squeamish about adding this option without any automated way to test it, but if you're okay with it, I can remove all of these changes to the testing file.

vangent commented on blob/azureblob/azureblob_test.go:
```diff
@@ -32,15 +32,17 @@ import (
 //
-// 4. Set the environment variables AZURE_STORAGE_ACCOUNT, AZURE_STORAGE_KEY to
-//    the storage account name and your access key.
+// 4. Set the environment variables AZURE_STORAGE_ACCOUNT, AZURE_STORAGE_KEY, and AZURE_CLOUD_ENVIRONMENT to
```
I think you can leave this unchanged; we're not going to `-record` in anything other than AzureCloud so we can just leave it empty.

vangent commented on blob/azureblob/azureblob_test.go:
```diff
@@ -113,7 +123,7 @@ func (h *harness) HTTPClient() *http.Client {
 
 func (h *harness) MakeDriver(ctx context.Context) (driver.Bucket, error) {
-	return openBucket(ctx, h.pipeline, accountName, bucketName, &Options{Credential: h.credential})
+	return openBucket(ctx, h.pipeline, accountName, bucketName, &Options{Credential: h.credential, CloudEnvironment: cloudEnv})
```
Same.

vangent commented on blob/azureblob/azureblob_test.go:
```diff
@@ -76,6 +79,13 @@ func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
 		if err != nil {
 			t.Fatal(err)
 		}
+		env, err := DefaultCloudEnvironment()
```
Same.

Strife96 commented:
PTAL: I went ahead and made the changes to address the comments from the first iteration. The test harness modifications have been reverted, and the environment variable is now used as a domain name rather than a mapped value. Because this now differs from other existing usages of "AZURE_CLOUD_ENVIRONMENT", the environment variable is now renamed to "AZURE_BLOB_DOMAIN". The default target remains "blob.core.windows.net". Some documentation was added on what values are expected and how to find them.

Over the course of these commits 3 test builds were run. Travis build 7400 succeeded, then build 7401 passed on all environments but OSX (https://travis-ci.com/google/go-cloud/builds/142083379). The failed tests were in the runtimevar/httpvar module. Between these two builds, only whitespace removals were made from running gofmt -w -s on azureblob.go and azureblob_test.go. After changing only documentation, Travis build 7402 then succeeded on all environments again. The error appears transient, but I am unsure of the specifics of runtimevar/httpvar and whether or not it is related to the changes in these commits.

vangent commented on blob/azureblob/azureblob.go:
```diff
@@ -253,6 +270,16 @@ func DefaultSASToken() (SASToken, error) {
+func DefaultCloudEnvironment() (CloudEnvironment, error) {
+	s := os.Getenv("AZURE_BLOB_DOMAIN")
+	if s == "" {
+		return CloudEnvironment("blob.core.windows.net"), nil // If unspecified, use default public domain.
```
Nit: this would read simpler with `s = "blob.core.windows.net"` instead of a separate `return`.

vangent replied:
Or, could just leave it the empty string, since `openBucket` handles the default.

vangent commented on blob/azureblob/azureblob.go:
```diff
@@ -223,6 +236,10 @@ type AccountKey string
 // https://docs.microsoft.com/en-us/azure/storage/common/storage-dotnet-shared-access-signature-part-1
 type SASToken string
 
+// CloudEnvironment is an Azure Cloud Environment domain name to target (i.e. AzureCloud, AzureUSGovernment, AzureChinaCloud).
```
Nit: the examples are stale.

vangent commented on blob/azureblob/azureblob.go:
```diff
@@ -223,6 +236,10 @@ type AccountKey string
 
+// CloudEnvironment is an Azure Cloud Environment domain name to target (i.e. AzureCloud, AzureUSGovernment, AzureChinaCloud).
+// It is read from the AZURE_BLOB_DOMAIN environment variable.
+type CloudEnvironment string
```
Naming: it's a bit weird to have the type and variable called `Environment` but the docstring and env variable called `Domain`. How about using `CloudEnvironmentDomain` everywhere? (e.g., `AZURE_BLOB_ENVIRONMENT_DOMAIN`, `CloudEnvironmentDomain` type, etc.).

vangent commented on blob/azureblob/azureblob.go:
```diff
@@ -26,6 +26,11 @@
 // The default URL opener will use credentials from the environment variables
 // AZURE_STORAGE_ACCOUNT, AZURE_STORAGE_KEY, and AZURE_STORAGE_SAS_TOKEN.
 // AZURE_STORAGE_ACCOUNT is required, along with one of the other two.
+// AZURE_BLOB_DOMAIN can optionally be used to provide an Azure Environment
```
Naming nit: let's use `AZURE_STORAGE_` prefix for consistency with the other ones.

Strife96 commented:
PTAL: Removed redundant default domain check, openBucket now exclusively handles the empty string case. Renamed the environment variable from AZURE_BLOB_DOMAIN to AZURE_STORAGE_DOMAIN and the internal type from CloudEnvironment to StorageDomain, as that is more accurate for its function now. Updated the documentation to reflect these changes and fix the outdated examples. 

Please let me know if I missed anything or if you would like any other changes.

vangent commented on blob/azureblob/azureblob_test.go:
```diff
@@ -371,18 +375,24 @@ func TestOpenerFromEnv(t *testing.T) {
 func TestOpenBucketFromURL(t *testing.T) {
 	prevAccount := os.Getenv("AZURE_STORAGE_ACCOUNT")
 	prevKey := os.Getenv("AZURE_STORAGE_KEY")
+	prevEnv := os.Getenv("AZURE_CLOUD_ENVIRONMENT")
```
One last thing -- please fix these env references.

Strife96 commented:
Oops, that's silly of me. Fixed!
